marahgh/assignment3/wordSearch.cpp

```diff
@@ -0,0 +1,212 @@
+  bool isEmpty() { return empty; }
+  /*returns the i'th row in the grid if the grid is not empty,and nullptr
+   * otherwise*/
+  char* operator[](int i) {
+    if (isEmpty()) {
+      return nullptr;
+    }
+    return matrix[i];
+  }
+};
+
+/*Dictionary class.
+Dictionary words are lowercase alphabet words*/
+class Dictionary {
+  vector<string> dictionary;
+
+  /****************************************************************************/
+  /* a private auxiliary method of isPrefix,
+  checks whether w2 is a prefix of w1*/
+  bool isWordPrefix(string w1, string w2) {
+    if (w1.size() < w2.size()) {
+      return false;
+    }
+    for (int i = 0; i < w2.size(); i++) {
+      if (w1[i] != w2[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /*Removes hyphens("-") and converts letters to lower case(returns a legal
+  dictionary word)
+  if the word contains alphabet letters and hyphens, other wise returns an empty
+  word */
+  string fixWord(string word) {
+    string fixed = "";
+    /*a loop to remove hyphens*/
+    for (char c : word) {
+      if (c != '-') {
+        fixed.append(1, tolower(c));
+      }
+    }
+    /*checks if the fixed word is dictionary legal word
+    (contains only lowercase alphabet letters)*/
+    for (char c : fixed) {
+      if (c < 'a' || c > 'z') {
+        return "";
+      }
+    }
+    return fixed;
+  }
+  /****************************************************************************/
+ public:
+  /*creates an empty dictionary*/
+  Dictionary() { dictionary = vector<string>(); }
+
+  /*Returns whether the given string is a valid word*/
+  bool isWord(string word) {
+    return isFoundInVector(dictionary, fixWord(word));
+  }
+
+  /*Adds a new non- empty word to the dictionary*/
+  bool addWord(string word) {
+    string fixed = fixWord(word);
+    if (isWord(fixed) || fixed.empty()) {
+      return false;
+    }
+    dictionary.push_back(fixed);
+    return true;
+  }
+
+  /*Returns whether the given string is a prefix of at least one word in the
+dictionary*/
+  bool isPrefix(string word) {
+    for (string w : dictionary) {
+      if (isWordPrefix(w, word)) {
+        return true;
+      }
+    }
+    return false;
+  }
+};
+
+/*An auxiliary function of find word, receives a grid,
+cell cordinates to start searchnig a word from, a dictionary.
+returns the legal words starts form cell i,j int the words vector*/
+void findWord(Grid& grid, int i, int j, Dictionary dictionary, string word,
+              vector<string>& words) {
+  if (i < 0 || j < 0 || i >= grid.getRows() || j >= grid.getCols() ||
+      grid[i][j] == '*' || grid.isEmpty()) {
+    return;
+  }
+  word.append(1, grid[i][j]);
+  if (!(dictionary.isPrefix(word))) {
+    word.pop_back();
+    return;
+  }
+  if (dictionary.isWord(word)) {
+    if (!isFoundInVector(words, word)) {
+      words.push_back(word);
+    }
+  }
+  char c = grid[i][j];
+  grid[i][j] = '*';

```

Comment:
It would be nicer to be able to do this without modifying `grid` along the way.
Generally modifying the data structure you are working on is not ideal: for example, if you wanted to search for words in different languages in parallel, this code would not work correctly as the two different runs would overwrite each other's grid.